File: cambio.py
#Cambio env
import random
from player import Player


class Cambio:
    # Define constants for special cards
    RED_KING_DIAMOND = 38
    RED_KING_HEART = 51
    JOKER_1 = 52
    JOKER_2 = 53
    DECK_SIZE = 54

    POWER_CARDS = Player.POWER_CARDS


    def __init__(self):
        self.deck = [i for i in range(54)]
        #Cambio has 52 cards + 2 jokers (ignore for now)

        random.shuffle(self.deck)
        #In the game of Cambio each player starts with 4 cards
        self.player_one = Player([self.deck.pop() for _ in range(4)], "Player 1")
        self.player_two = Player([self.deck.pop() for _ in range(4)], "Player 2")

        self.discard_pile = []
        self.turn_count = 0
        self.game_over = False
        self.current_player_turn = 1
        
    def starting_peek(self):
        self.player_one.peek_self()
        self.player_two.peek_self()
        
    
    def get_game_details(self):
        return (self.player_one.get_inventory(), self.player_two.get_inventory())

    def get_deck(self):
        return self.deck

    def get_player(self, player = 1):
        if player == 1:
            return self.player_one.get_inventory()
        else:
            return self.player_two.get_inventory()

    def convert_card(self,card):
        #check for -2
        if card == -2:
            return "No Card"
        #check for joker
        if card >= 52:
            return "JK"
        #turn the int values into a card
        cards = ["A","2","3","4","5","6","7","8","9","10","J","Q","K","JK"][card%13]
        #A = 1, JK = joker
        #// MEANS DIVIDE AND FLOOR (FLOOR REMOVES DECIMALS)
        suits = 'SCDH'[card//13]
        return cards + suits
    
    def get_card_score(self,card):
        if card >= 52:
            return 0
        #In cambio red kings are minus one so add check here
        if card == self.RED_KING_DIAMOND or card == self.RED_KING_HEART:
            #red kings are worth -1
            return -1
        score = card % 13
        score = score + 1
        if score >= 10:
            return 10

        return score
    
    def discard(self,player = 1):
        current_player = self.player_one if player == 1 else self.player_two
        card = current_player.get_in_hand()
        if card != -2:
            self.discard_pile.append(card)
            current_player.set_in_hand(-2)
    
    def player_put_card_in_hand_into_deck(self, hand_index, player = 1):
        current_player = self.player_one if player == 1 else self.player_two
        current_player.swap_hand_with_inventory(hand_index)
        self.discard(player)
    
    def discard_card_from_hand(self, player = 1):
        current_player = self.player_one if player == 1 else self.player_two
        if current_player.get_in_hand() == -2:
            raise Exception("No card in hand to discard")
        self.discard(player)

    def player_get_card_from_pile(self,player = 1):
        current_player = self.player_one if player == 1 else self.player_two
        #quick check to see if player has a card in hand already
        if current_player.get_in_hand() != -2:
            raise Exception(f"Player {player} already has a card in hand")
        if len(self.deck) == 0:
            raise Exception("Deck is empty")
        card = self.deck.pop()
        current_player.set_in_hand(card)

    def step(self):

        if len(self.deck) == 0:  
            self.game_over = True
            return self.get_winner()
        else:
            #simulate one turn of the game
            self.turn_count += 1
            self.last_drawn = None
            self.last_power = None

            current_player = self.player_one if self.current_player_turn == 1 else self.player_two
            player_id = self.current_player_turn

            # Player draws a card
            self.player_get_card_from_pile(player_id)
            self.last_drawn = current_player.get_in_hand()

            # If card is a power card, use it and end the turn
            power = self.use_power(current_player.get_in_hand(), player_id)
            if power:
                self.last_power = power
                self.discard(player_id)
                self.current_player_turn = 2 if self.current_player_turn == 1 else 1
                return

            # Player decides what to do (swap or discard)
            swap_index = current_player.decide_swap_index()
            if swap_index != -1:
                self.player_put_card_in_hand_into_deck(swap_index, player_id)

            # Discard if they still have a card in hand
            if current_player.get_in_hand() != -2:
                self.discard(player_id)

            self.current_player_turn = 2 if self.current_player_turn == 1 else 1
            
    def get_card_rank(self,card):
        #returns a string value for the card rank
        return ["A","2","3","4","5","6","7","8","9","10","J","Q","K","JK"][card%13]


    def turn_deck_to_name(self,player = 1):
        deck = self.player_one.get_inventory() if player == 1 else self.player_two.get_inventory()
        cards = []
        for card in deck:
            cards.append(self.convert_card(card))
        
        return cards
    
    def turn_deck_to_score(self, player = 1):
        deck = self.player_one.get_inventory() if player == 1 else self.player_two.get_inventory()
        score = 0
        for card in deck:
            #checks each card in the player one deck and turns it into a value
            score += self.get_card_score(card)
        
        return score
    
    
    def get_winner(self):
        p1_score = self.turn_deck_to_score()
        p2_score = self.turn_deck_to_score(2)

        if p1_score == p2_score:
            return "--- DRAW ---"
        elif p1_score > p2_score:
            return "--- P2 Wins ---"
        else:
            return "--- P1 Wins ---"
        
    def get_discard_pile(self):
        return self.discard_pile
    
    def use_power(self,card,player = 1):
        #If the card in the player hand is not a power card then return False
        player = self.player_one if player == 1 else self.player_two
        power = player.get_power(card)
        if not power:
            return False
        if power == "PEEK_SELF":
            player.peek_self()
        elif power == "PEEK_OPPONENT":
            player.peek_opponent(self.player_two.get_inventory())
        elif power == "BLIND_SWAP":
            self.swap_player_cards(*player.decide_blind_swap())

        return power

    def swap_player_cards(self, p1_index, p2_index):
        if p1_index == -1 or p2_index == -1:
            return
        inv1 = self.player_one.get_inventory()
        inv2 = self.player_two.get_inventory()
        inv1[p1_index], inv2[p2_index] = inv2[p2_index], inv1[p1_index]

        
        

File: cplusplus_env/cambio.cpp
#include "cambio.h"
#include <algorithm>
#include <random>
#include <stdexcept>
#include <iostream>

const int Cambio::RED_KING_DIAMOND;
const int Cambio::RED_KING_HEART;
const int Cambio::JOKER_1;
const int Cambio::JOKER_2;
const int Cambio::DECK_SIZE;

const std::unordered_map<int, std::string> Cambio::POWER_CARDS = {
    {6,  "PEEK_SELF"},
    {7,  "PEEK_SELF"},
    {8,  "PEEK_OPPONENT"},
    {9,  "PEEK_OPPONENT"},
    {10, "BLIND_SWAP"},
    {11, "SINGLE_PEEK_SWAP"},
    {12, "DOUBLE_PEEK_SWAP"}
};

Cambio::Cambio()
    : deck(),
      discard_pile(),
      player_one({}, "Player 1"),
      player_two({}, "Player 2"),
      turn_count(0),
      game_over(false),
      current_player_turn(1) {

    for (int i = 0; i < 54; i++) deck.push_back(i);

    std::random_device rd;
    std::mt19937 g(rd());
    std::shuffle(deck.begin(), deck.end(), g);

    std::vector<int> p1_hand, p2_hand;
    for (int i = 0; i < 4; i++) {
        p1_hand.push_back(deck.back()); deck.pop_back();
        p2_hand.push_back(deck.back()); deck.pop_back();
    }
    player_one.set_inventory(p1_hand);
    player_two.set_inventory(p2_hand);
}

// Getters
std::vector<int> Cambio::get_deck() {
    return deck;
}

std::vector<int> Cambio::get_discard_pile() {
    return discard_pile;
}

std::vector<int> Cambio::get_player(int player) {
    if (player == 1) return player_one.get_inventory();
    return player_two.get_inventory();
}

int Cambio::get_turn_count() {
    return turn_count;
}

bool Cambio::get_game_over() {
    return game_over;
}

// Logic methods
int Cambio::get_card_score(int card) {
    if (card >= 52) return 0;
    if (card == RED_KING_DIAMOND || card == RED_KING_HEART) return -1;
    int score = (card % 13) + 1;
    return score >= 10 ? 10 : score;
}

std::string Cambio::convert_card(int card) {
    if (card == -2) return "No Card";
    if (card >= 52) return "JK";
    std::string cards[] = {"A","2","3","4","5","6","7","8","9","10","J","Q","K","JK"};
    std::string suits = "SCDH";
    return cards[card % 13] + suits[card / 13];
}

void Cambio::discard(int player) {
    Player& current = (player == 1) ? player_one : player_two;
    int card = current.get_in_hand();
    if (card != -2) {
        discard_pile.push_back(card);
        current.set_in_hand(-2);
    }
}

void Cambio::discard_card_from_hand(int player) {
    Player& current = (player == 1) ? player_one : player_two;
    if (current.get_in_hand() == -2)
        throw std::runtime_error("No card in hand to discard");
    discard(player);
}

void Cambio::player_get_card_from_pile(int player) {
    Player& current = (player == 1) ? player_one : player_two;
    if (current.get_in_hand() != -2)
        throw std::runtime_error("Player already has a card in hand");
    int card = deck.back();
    deck.pop_back();
    std::cout << "Player " << player << " drew " << convert_card(card) << std::endl;
    current.set_in_hand(card);
}

void Cambio::player_put_card_in_hand_into_deck(int hand_index, int player) {
    Player& current = (player == 1) ? player_one : player_two;
    current.swap_hand_with_inventory(hand_index);
    discard(player);
}

int Cambio::turn_deck_to_score(int player) {
    std::vector<int> inv = (player == 1) ? player_one.get_inventory() : player_two.get_inventory();
    int score = 0;
    for (int card : inv) score += get_card_score(card);
    return score;
}

std::vector<std::string> Cambio::turn_deck_to_name(int player) {
    std::vector<int> inv = (player == 1) ? player_one.get_inventory() : player_two.get_inventory();
    std::vector<std::string> names;
    for (int card : inv) names.push_back(convert_card(card));
    return names;
}

std::string Cambio::get_winner() {
    int p1 = turn_deck_to_score(1);
    int p2 = turn_deck_to_score(2);
    if (p1 == p2) return "--- DRAW ---";
    else if (p1 > p2) return "--- P1 Loses ---";
    else return "--- P1 Wins ---";
}

std::string Cambio::step() {
    if (deck.empty()) {
        game_over = true;
        return get_winner();
    }

    turn_count++;
    int player_id = current_player_turn;

    player_get_card_from_pile(player_id);

    Player& current = (player_id == 1) ? player_one : player_two;
    int swap_index = current.decide_swap_index();
    if (swap_index != -1) {
        player_put_card_in_hand_into_deck(swap_index, player_id);
    }

    if (current.get_in_hand() != -2) {
        discard(player_id);
    }

    current_player_turn = (current_player_turn == 1) ? 2 : 1;

    return "";
}
